boards/Makefile.common

```diff
@@ -381,7 +352,7 @@ debug-lst:  $(TARGET_PATH)/debug/$(PLATFORM).lst
 doc: | target
 	@# This mess is all to work around rustdoc giving no way to return an
 	@# error if there are warnings. This effectively simulates that.
-	$(Q)RUSTDOCFLAGS='$(RUSTDOC_FLAGS_TOCK)' $(CARGO) --color=always doc $(VERBOSE_FLAGS) --release --package $(PLATFORM) --target-dir=$(TARGET_DIRECTORY) 2>&1 | grep -C 9999 warning && (echo "Warnings detected during doc build" && if [[ $$NOWARNINGS == "true" ]]; then echo "Erroring due to CI context" && exit 33; fi) || if [ $$? -eq 33 ]; then exit 1; fi
+	$(Q)RUSTDOCFLAGS='$(RUSTDOC_FLAGS_TOCK)' $(CARGO) --color=always doc $(VERBOSE_FLAGS) --release --package $(PLATFORM) --target-dir=$(TARGET_DIRECTORY) --target x86-64-unknown-linux-gnu 2>&1 | grep -C 9999 warning && (echo "Warnings detected during doc build" && if [[ $$NOWARNINGS == "true" ]]; then echo "Erroring due to CI context" && exit 33; fi) || if [ $$? -eq 33 ]; then exit 1; fi
```

Comment:
This will break on non-x86-64 Linux systems, as they won't have that target installed by default. We'll need some way to ask `rustc` about its default target and inject it here.

Reply:
So, the _actual_ right solution is to document for the actual target. The reason this needs to be here is because, by default, cargo will now compile for the board target and it turns out there are a bunch of bugs in the documentation (broken links etc) such that it breaks the build. The _bug_ is that this documentation is broken, and using an explicit host target here is a band-aid.

Reply:
I think, unfortunately, that means we have to fix the docs.

It's unfortunate how broken all of this is in the cargo/rustc world, and unfortunate you happened to find the issue, but I don't think we can ignore it now.

Reply:
> So, the _actual_ right solution is to document for the actual target.

You can document anything that has one unique target for that given target, but running `cargo doc` for any subsequent target will re-generate docs for common imported crates (e.g., `kernel`), which again can include target-specific code or docs.

The real underlying issue is that rustdoc isn't entirely independent of targets and lacks (complete) semantic understanding of what it means to compile for multiple targets in the presence of cfg-attributes, etc. It will not go down all code paths of all targets, but render everything for a particular target and configs only.

We could fix this by running `cargo doc` from all top-level boards and copying the entire `target/doc` folder for each board, but that would be very confusing and waste space.

Ultimately, this is not a major concern for us in practice, as most generic crates (`kernel`, `capsules`, ...) do not generally feature any conditional compilation -- and as such, the docs for them should look identical. For those common crates though, we should use a rustdoc target that the user has installed, though, and that target is not `x86-64-unknown-linux-gnu` (e.g., macOS users will not have this installed by default). We might use one of the targets we support anyways, or the host's target which rustup installs by default.

Reply:
> We could fix this by running `cargo doc` from all top-level boards and copying the entire `target/doc` folder for each board, but that would be very confusing and waste space.

This is exactly what we do in CI.

Reply:
But if we are buying into cargo, then don't we have to buy into cargo and do whatever they think should happen when running `cargo doc`?

Reply:
Not quite: https://github.com/tock/tock/blob/c056bec40da4bb97ad498248bc5b83896b51a710/Makefile#L151-L156

This overwrites common crates, as all of these `make doc` invocations write to `target/doc` subdirectories. An ARM doc build of the kernel crate will overwrite a RISC-V one.

Long story short: this kind of sucks, and I agree that specifying a target is a temporary band-aid for what needs to be a proper fix of our docs. But if something doesn't change in Rustdoc (or we don't _actually_ copy all crate's doc folders after each build), we'll never have actual consistent target-specific, or entirely target-independent docs. It'll always be a wild mix of targets depending on the order in which we run these `make doc` commands.

Reply:
I think the fix is easy. Two small patches, and then just remove this change. We get better docs, and, since there would be no change on this file, we would go back to just not caring about what exactly happens.

Before noticing this problem:
<img width="1230" alt="image" src="https://github.com/tock/tock/assets/1467890/62768fe3-78a3-4828-97d5-22e4e883d805">
After noticing this problem:
<img width="1199" alt="image" src="https://github.com/tock/tock/assets/1467890/19266915-dda2-4f76-935c-dee19d6a88c4">

Reply:
Need #4057 and #4059.

Reply:
> do whatever they think should happen when running `cargo doc`?

They think that we should fix our conditional compiles to always include `doc` in the target of whichever "real" thing is:
https://github.com/encapfn/encapfn/blob/825b6ff0a5ce840a89edd4ce03c79d47f5d39b07/src/rt/mock/stack_alloc.rs#L9-L10

As for cases where having multiple `cfg`s that would clash if they both had `doc` passed, I have no idea. This is yet another win for Tock's avoidance of conditional compilation, which doesn't make this a huge deal for us in practice. It's just not great that with this change, we'd be building docs for some crates for an arbitrary arch.

Reply:
Honestly, I forget that it's even possible to compile a crate entirely differently based on what architecture you are compiling for. In my opinion, if we are doing that to the point where the docs change we deserve to have broken documentation because how is anyone supposed to use our libraries anyway.

In any case, we have built a sane design where this isn't a problem, so, let's just ignore this oddity.